### ISSUE TITLE
Regression in code generation of cached nodes

### ISSUE THREAD

**Issue by spicyj** [babel/babel#2078]
(Not sure if the issue title is actually accurate, please rename.)

Previously

```js
var babel = require('./packages/babel');

var plugin = new babel.Transformer('test', {
  JSXElement: function (node, parent, scope, file) {
    var openName = node.openingElement.name.name;
    if (openName === 'foo' && node.children[0].type === 'Literal') {
      return node.children[0];
    }
  }
});

console.log(babel.transform('var x = <foo>bar</foo>;', {whitelist: [], plugins: [plugin]}).code);
```

printed 

```
var x = "bar";
```

but now it prints

```
var x = bar;
```

I bisected this to f0994f106eb5270508bf523601a827fcf77f6fe5. I assume the old printing of the node is getting cached somehow. Bug or feature?

**Comment by sebmck** [comment]
Just a consequence of how the JSX content nodes are constructed. Ideally the contents shouldn't even be represented with a `Literal` node but with a `JSXLiteral` node but that'd be a breaking change to the AST. Temp fix is to do `node.children[0].rawValue = null`;

**Comment by sebmck** [comment]
Drives me crazy how much the `Literal` node has been overloaded, it represents all null values, numbers, strings, regexes and JSX inner text. Really hope to change in this in the future as it means a lot of transforms are much more brittle than they should be.

**Comment by sebmck** [comment]
Wait nope, that wont do it. `<div>"foo"</div>` will return a false positive. What a mess.

**Comment by spicyj** [comment]
Everything in that list except JSX text sounds sane enough to me…

**Comment by sebmck** [comment]
I'm trying to push a new release but some of the tests are timing out because my current laptop is extremely terrible and slow. Will try it in a bit once it calms down.

> Everything in that list except JSX text sounds sane enough to me…

They're all separate syntactic constructs and representing them with the exact same node is a bit silly. You end up with not being able to use it without first checking the type of `node.value`. It gets even more tricky with regexes since, if you're in an environment that doesn't support for example, the unicode flag on a regex, `node.value` will be `undefined` and so you have to use `node.regex.*`.

**Comment by kevinb7** [comment]
@spicyj thanks for reporting this.  @sebmck thanks for the quick fix.

As for overloading of "Literal" nodes, it also erases information, e.g. `{ type: "Literal", value: 3.0 }`.  When you go to codegen this you end up with `3` instead of `3.0` which is not what you want when all you're doing is fixing formatting.   So then you have attach a `raw` field which seems kind of silly when something like `{ type: "NumberLiteral", value: "3.0" }` would've done the trick.

**Comment by sebmck** [comment]
@kevinb7 

> As for overloading of "Literal" nodes, it also erases information,

An AST doesn't need to have any CST information.

>  So then you have attach a `raw` field which seems kind of silly when something like `{ type: "NumberLiteral", value: "3.0" }` would've done the trick.

That makes it extremely awkward to deal with number nodes then since you need to parse the number whenever you want to deal with it and when you're dealing with a number node you almost always want to.

**Comment by spicyj** [comment]
> It gets even more tricky with regexes since, if you're in an environment that doesn't support for example, the unicode flag on a regex, `node.value` will be `undefined` and so you have to use `node.regex.*`.

That's totally a fair complaint.

**Comment by kevinb7** [comment]
> An AST doesn't need to have any CST information.

Good point.  I'm happy to see people are starting to talk about CSTs.  ASTs and CSTs each have their purpose.  I can definitely see how having to parse literals when doing AST transforms would be a pain.

**Comment by sebmck** [comment]
Published 5.8.6, thanks!

**Comment by spicyj** [comment]
Thank you!
